README.md
[![Python 3](https://img.shields.io/badge/python-3-blue.svg)](https://www.python.org/) 

CityGML変換ツール (citygml-convert-tools)
===

[CityGML](https://www.mlit.go.jp/plateau/learning/)は、地理空間データのため標準データフォーマットです。
XMLベースで定義されているCityGMLは、中間データフォーマットと言われているように利用サイドが使いやすい形式に変換して利用することを想定されている。
手軽にCityGMLファイルを変換できるツールを作成した

コードは出来るだけシンプルにしたかったので、サンプルコードとしてバッサリとコードは簡潔にしてます。  
エラー処理やテスト等も省略している部分があります。何かの参考にしていただければ幸いです  

### about py_plateau module

```
$ python cct.py --version     

CityGML convert tools (CCT) v0.0.1
```

#### Features

- リポジトリルートに配置されているcct.pyを利用することで、py_plateauモジュールを利用したCityGMLの変換が可能
- 以下の機能が利用可能
  - CityGMLファイル → PLYファイル変換
  - CityGMLファイル → GeoJSONファイル変換
  - 建物(bldg)のみ対応
  - LOD0, LOD1, LOD2対応
  - テクスチャは非対応
  - [緯度経度]、[経度緯度]の変換対応
  - 建物(bldg)分割

※1 2D座標系に変換する場合も3D座標で[longitude, latitude, height]に変換します

#### Usage

##### Main

```
$ python cct.py --help

Usage: cct.py [OPTIONS] COMMAND [ARGS]...

  CityGML convert tools (CCT) v0.0.1

Options:
  --version                 Show the version and exit.
  -d, --debug / --no-debug  debug mode
  -v, --verbose             verbose mode
  --help                    Show this message and exit.

Commands:
  geojson  Convert CityGML file to GeoJSON file
  ply      Convert CityGML file to PLY file
```

##### CityGMLファイル → PLYファイル変換 
```
$ python cct.py ply --help

Usage: cct.py ply [OPTIONS] FILENAME

  Convert CityGML file to PLY file

Options:
  -o, --output TEXT        output path name
  -s, --to-srid TEXT       output SRID(EPSG)
  -l, --lod INTEGER RANGE  output lod type  [0<=x<=2]
  -sp, --separate          separate the building data
  --help                   Show this message and exit.
```

Examples
```
$ python cct.py ply 53392633_bldg_6697_2_op.gml --lod=2 --to-srid=6677
```

##### CityGMLファイル → GeoJSONファイル変換 
```
$ python cct.py geojson --help

Usage: cct.py geojson [OPTIONS] FILENAME

  Convert CityGML file to GeoJSON file

Options:
  -o, --output TEXT        output path name
  -s, --to-srid TEXT       output SRID(EPSG)
  -l, --lod INTEGER RANGE  output lod type  [0<=x<=2]
  -sp, --separate          separate the building data
  -lonlat, --lonlat        swap lon lat order
  --help                   Show this message and exit.
```

Examples
```
$ python cct.py geojson 53392633_bldg_6697_2_op.gml --lod=2 --to-srid=4326
```

Installation
---
Pythonと[poetry](https://python-poetry.org/)を利用します。

```
$ git clone git@github.com:GeoPythonJP/citygml-convert-tools.git
$ poetry install
$ poetry shell
```

Changelog
---
see the [changelog](./CHANGELOG.md).


Tests Data
---
[3D都市モデル（Project PLATEAU）東京都23区](https://www.geospatial.jp/ckan/dataset/plateau-tokyo23ku)のCityGMLの東京都大田区羽田空港三丁目データ ”53392633_bldg_6697_2_op.gml” のみで動作確認

License
---
MIT
See the[license](./LICENSE) document for the full text.

Modules
---
下記のモジュールを参考、使用しています。
各々のライセンスに従ってください。
* [earcut-python](https://github.com/joshuaskelly/earcut-python)
  * earcut.py
* [AcculusSasao/plateaupy](https://github.com/AcculusSasao/plateaupy)
* [ksasao/PlateauCityGmlSharp](https://github.com/ksasao/PlateauCityGmlSharp/)
* [cityjson/cjio](https://github.com/cityjson/cjio)


Contributors
---
| GitHub                                          | Twitter                                             | 
|-------------------------------------------------|-----------------------------------------------------| 
| [homata](http://github.com/homata)              | [@homata](https://twitter.com/homata)               |
| [nokonoko1203](https://github.com/nokonoko1203) | [@nokonoko_1203](https://twitter.com/nokonoko_1203) |

CityGMLビューワー
---
CityGMLファイルを表示させる下記のツールがあります。

* Windows版
    * [FZKViewer](https://www.iai.kit.edu/1302.php)
        * [FZKViewer のインストール（Windows 上）](https://www.kkaneko.jp/tools/win/fzkviewer.html)
* Mac版
    * [azul](https://github.com/tudelft3d/azul)

座標系のメモ
---
座標系の変換等をする場合の参考情報を記述します

#### CityGMLの座標系
* 日本測地系2011における経緯度座標系と東京湾平均海面を基準とする標高の複合座標参照系: 6697
* 日本測地系2011における経緯度座標系: 6668

#### 日本の測地系
* [日本の測地系](https://www.gsi.go.jp/sokuchikijun/datum-main.html)
* [わかりやすい平面直角座標系](https://www.gsi.go.jp/sokuchikijun/jpc.html)

#### EPSG:6697
EPSG:6697というのは「JGD2011 + JGD2011 (vertical) height」という座標参照系でEPSG:6668（JGD2011）とEPSG:6695（JGD2011 (vertical) height）からなる座標系である

* [EPSG:6697](https://epsg.io/6697): JGD2011 + JGD2011 (vertical) height　緯度、経度、標高 (EPSG:6668+6695)
  * [EPSG:6668](https://epsg.io/6668): JGD2011　緯度、経度
  * [EPSG:6695](https://epsg.io/6695): JGD2011 (vertical) height　標高単位：m
* [EPSG:6666](https://epsg.io/6666):  JGD2011 地心直交座標系 (X, Y, Z) 地球中心のXYZ 単位：m

#### 日本測地系2011　平面直角座標系:6669-6687
```
系	EPSG (JGD2011)	区域
1	6669	長崎県、鹿児島県の一部
2	6670	福岡県、佐賀県、熊本県、大分県、宮崎県、鹿児島県の一部
3	6671	山口県、島根県、広島県
4	6672	香川県、愛媛県、徳島県、高知県
5	6673	兵庫県、鳥取県、岡山県
6	6674	京都府、大阪府、福井県、滋賀県、三重県、奈良県、和歌山県
7	6675	石川県、富山県、岐阜県、愛知県
8	6676	新潟県、長野県、山梨県、静岡県
9	6677	東京都の一部、福島県、栃木県、茨城県、埼玉県、千葉県、群馬県、神祭川県
10	6678	青森県、秋田県、山形県、岩手県、宮城県
11	6679	北海道の一部
12	6680	北海道の一部
13	6681	北海道の一部
14	6682	東京都の一部
15	6683	沖縄県の一部
16	6684	沖縄県の一部
17	6685	沖縄県の一部
18	6686	東京都の一部
19	6687	東京都の一部
```


PLATEAU関連情報
---
PLATEAU関連情報のメモ

* [Project PLATEAU](https://www.mlit.go.jp/plateau/)
    * [3D都市モデル標準製品仕様書（第2.3版）](https://www.mlit.go.jp/plateau/file/libraries/doc/plateau_doc_0001_ver02.pdf)
    * [3D都市モデル標準製品仕様書（第2.3版）(HTML)](https://www.mlit.go.jp/plateaudocument/)
    * [3D都市モデル整備のためのBIM活用マニュアル（第1.0版）](https://www.mlit.go.jp/plateau/file/libraries/doc/plateau_doc_0003_ver01.pdf)
    * [3D都市モデルのデータ変換マニュアル](https://www.mlit.go.jp/plateau/file/libraries/doc/plateau_doc_0007_ver01.pdf)
        * [3D都市モデルのデータ変換マニュアル](https://github.com/Project-PLATEAU/Data-Conversion-Manual-for-3D-City-Model)
    * [G空間情報センター](https://www.geospatial.jp/ckan/dataset/plateau)
    * [GitHub](https://github.com/Project-PLATEAU)
* [i－都市再生技術仕様（案）／i-UR Technical Materials](https://www.chisou.go.jp/tiiki/toshisaisei/itoshisaisei/iur/index.html)
* [東京都デジタルツイン実現プロジェクト](https://info.tokyo-digitaltwin.metro.tokyo.lg.jp/)
    * [CityJSONに変換するcitygml-toolsの日本語マニュアル](https://github.com/tokyo-digitaltwin/citygml-tools)
* [国土交通データプラットフォーム](https://www.mlit-data.jp/platform/)
* [スマートシティ官民連携プラットフォーム](https://www.mlit.go.jp/scpf/)

### 仕様と変換ツール類
* [CityGML](https://www.ogc.org/standards/citygml)
* [CityJSON](https://www.cityjson.org/)
* [FME](https://www.safe.com/fme/)
* [3dcitydb/3dcitydb](https://github.com/3dcitydb/3dcitydb)

### CityGML
* [CityGML 3.0のGML仕様のドラフト](https://www.ogc.org/standards/requests/257)

### i-UR
* i-UR1.4は名前空間及びXMLSchemaファイルの所在が変更されたことに伴い、i-UR1.5に改定されている (https://www.chisou.go.jp/tiiki/toshisaisei/itoshisaisei/iur)

### 3dcitydb
TBD

### CityJSON
TBD

### OpenStreetMap
* [JA:MLIT PLATEAU/imports outline](https://wiki.openstreetmap.org/wiki/JA:MLIT_PLATEAU/imports_outline)
  * [議論：JA talk:MLIT PLATEAU/imports outline](https://wiki.openstreetmap.org/wiki/JA_talk:MLIT_PLATEAU/imports_outline#既存データを編集していたマッパーへの、OSMメッセージ連絡の要否)
* [JA talk:MLIT PLATEAU](https://wiki.openstreetmap.org/wiki/JA_talk:MLIT_PLATEAU)
  * [citygml-osm](https://github.com/yuuhayashi/citygml-osm)
    * [citygml-osm wiki](https://github.com/yuuhayashi/citygml-osm/wiki/Transformation)
* [Plateau建物データ: OpenStreetMapへのインポート手順（ドラフト）](https://qiita.com/nyampire/items/1c10afdd36750c87154d)

### 便利なツール
* [CloudCompare](https://cloudcompare.org/)
* [Meshlab](https://www.meshlab.net/)
* [Blender.jp](https://blender.jp/)
* [Cesium](https://cesium.com/)
* [QGIS](https://qgis.org/)

### その他

関連情報のメモ
* [Geo関連情報のメモ](./GEOMEMO.md)

### 3D都市空間モデルビューワー
3D都市空間モデルビューワーのサンプルです

* [3D都市空間モデルビューワー](./3dmodel_viewer/READMD.md)


CHANGELOG.md
# Changelog
In this changelog all modifications to the `citygml-convert-tools`

## 2022-11-23 
### Added
* 3D都市空間モデルビューワーを追加
* テクスチャーのコードを追加

### Fixed
* none

## 2022-11-13 (v0.0.1)
### Added
* コマンドラインをサブコマンド形式に変更
* PLYとGeoJSONの変換処理を統合
* PLYとGeoJSONのテストを追加
* earcut-pythonのサブモジュールを廃止し、プロジェクトに統合

### Fixed
* PLY変換のバグを修正

## 2022-11-05
### Added
* citygml2geojsonを追加

### Fixed
* none


cct.py
#!/usr/bin/env python
# coding: utf-8

import traceback

import click

import py_plateau
from py_plateau.city_gml import CityGml, Subset


@click.group(help=f"CityGML convert tools (CCT) v{py_plateau.__version__}")
@click.version_option(version=py_plateau.__version__, message="CityGML convert tools (CCT) v%(version)s")
@click.option("-d", "--debug/--no-debug", default=False, help="debug mode")
@click.option("-v", "--verbose", default=False, is_flag=True, help="verbose mode")
@click.pass_context
def main(context, debug, verbose):
    context.obj = dict(debug=debug, verbose=verbose)


@main.command(help="Convert CityGML file to GeoJSON file")
@click.argument("filename", type=click.Path(exists=True))  # input CityGML file name
@click.option("-o", "--output", "output_path", default="output", help="output path name")
@click.option("-s", "--to-srid", "to_srid", default="4326", help="output SRID(EPSG)")
@click.option("-l", "--lod", "lod", default=2, type=click.IntRange(0, 2), help="output lod type")
@click.option("-sp", "--separate", "separate", default=False, is_flag=True, help="separate the building data")
@click.option("-lonlat", "--lonlat", "lonlat", default=True, is_flag=True, help="swap lon lat order")
@click.pass_context
def geojson(context, filename, output_path, to_srid, lod, separate, lonlat):
    """Convert CityGML file to GeoJSON file"""

    try:
        if context.obj["verbose"]:
            click.echo(f"\nConvert CityGML file to GeoJSON file\n")
            click.echo(f" Options:")
            click.echo(f'  debug={context.obj["debug"]}')
            click.echo(f'  verbose={context.obj["verbose"]}')
            click.echo(f"  filename={filename}")
            click.echo(f"  output_path={output_path}")
            click.echo(f"  to_srid={to_srid}")
            click.echo(f"  lod={lod}")
            click.echo(f"  separate={separate}")
            click.echo(f"  lonlat={lonlat}")
            click.echo(f"\n")

        obj_city_gml = CityGml(filename, Subset.GEOJSON, to_srid, separate=separate, lonlat=lonlat)
        if lod == 0:
            obj_city_gml.lod0()
        elif lod == 1:
            obj_city_gml.lod1()
        elif lod == 2:
            obj_city_gml.lod2()
        else:
            raise Exception(f"ERROR: lod number = {lod}")

        obj_city_gml.write_file(output_path)

    except Exception as e:
        click.echo(e)
        traceback.print_exc()


@main.command(help="Convert CityGML file to PLY file")
@click.argument("filename", type=click.Path(exists=True))  # input CityGML file name
@click.option("-o", "--output", "output_path", default="output", help="output path name")
@click.option("-s", "--to-srid", "to_srid", default="6677", help="output SRID(EPSG)")
@click.option("-l", "--lod", "lod", default=2, type=click.IntRange(0, 2), help="output lod type")
@click.option("-sp", "--separate", "separate", default=False, is_flag=True, help="separate the building data")
@click.pass_context
def ply(context, filename, output_path, to_srid, lod, separate):
    """Convert CityGML file to PLY file"""
    try:
        if context.obj["verbose"]:
            click.echo(f"\nConvert CityGML file to PLY file\n")
            click.echo(f" Options:")
            click.echo(f'  debug={context.obj["debug"]}')
            click.echo(f'  verbose={context.obj["verbose"]}')
            click.echo(f"  filename={filename}")
            click.echo(f"  output_path={output_path}")
            click.echo(f"  to_srid={to_srid}")
            click.echo(f"  lod={lod}")
            click.echo(f"  separate={separate}")
            click.echo(f"\n")

        obj_city_gml = CityGml(filename, Subset.PLY, to_srid, separate=separate)
        if lod == 0:
            obj_city_gml.lod0()
        elif lod == 1:
            obj_city_gml.lod1()
        elif lod == 2:
            obj_city_gml.lod2()
        else:
            raise Exception(f"ERROR: lod number = {lod}")

        obj_city_gml.write_file(output_path)

    except Exception as e:
        click.echo(e)
        traceback.print_exc()


if __name__ == "__main__":
    main()
